include/quic/datagram.hpp

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include "utils.hpp"
+
+namespace oxen::quic
+{
+    class IOChannel;
+    class Connection;
+    class Endpoint;
+    class Stream;
+
+    // IO callbacks
+    using io_data_callback = std::function<void(IOChannel&, bstring_view)>;
+    using io_close_callback = std::function<void(IOChannel&, uint64_t error_code)>;
+    // returns 0 on success
+    using io_open_callback = std::function<uint64_t(IOChannel&)>;
+    using io_unblocked_callback = std::function<bool(IOChannel&)>;
+
+    class IOChannel
+    {
+      public:
+        IOChannel(Connection& c, Endpoint& e, io_data_callback data_cb = nullptr, io_close_callback close_cb = nullptr);
+
+        virtual ~IOChannel() { log::trace(log_cat, "{} called", __PRETTY_FUNCTION__); };
+
+        Connection& conn;
+        Endpoint& endpoint;
+        io_data_callback data_callback;
+        io_close_callback close_callback;
+        buffer_que user_buffers;
+
+        // no copy, no move. always hold in a shared pointer
+        IOChannel(const IOChannel&) = delete;
+        IOChannel& operator=(const IOChannel&) = delete;
+        IOChannel(IOChannel&&) = delete;
+        IOChannel& operator=(IOChannel&&) = delete;
+
+        virtual bool is_stream() = 0;
+        virtual bool is_empty() const = 0;
+        virtual std::shared_ptr<Stream> get_stream() = 0;
+        virtual void send(bstring_view data, std::shared_ptr<void> keep_alive = nullptr) = 0;
+        virtual std::vector<ngtcp2_vec> pending() = 0;
+        virtual void append_buffer(bstring_view buffer, std::shared_ptr<void> keep_alive) = 0;
+        virtual int64_t stream_id() const = 0;
+        virtual bool is_closing() const = 0;
+        virtual bool sent_fin() const = 0;
+        virtual void set_fin(bool v) = 0;
+        virtual size_t unsent() const = 0;
+        virtual void wrote(size_t n) = 0;
+    };
```

Comment:
This class is useful internally for managing the buffers, but it feels like this internal implementation detail is leaking out to the libquic-using application and forcing complications on *outside* code to make *inside* code more maintainable.

From an application's point of view, I'm not sure what the value is in having Stream and DatagramIO inherit from a base type (and then giving me the base type): that just means the application has to untangle the object structure to get at the thing they care about (which will be a stream *or* a datagram, but rarely or perhaps never both).  And while we can duplicate most of Stream's public interface into IOChannel (as is done here) that is an awkward API for the caller when using datagrams because we are going through a base class that has a bunch of methods that are meaningless for a datagram channel.

If you were to use the base class *internally* but still have the user use Stream and DatagramIO, we would get more manageable code internally but also more manageable code externally, which seems like a win.

Reply:
That's a really good point that I didn't consider. I guess I was trying to do 3 different things at the same time, and it "almost" worked
1) introduce the base IOChannel class and have datagram and stream inherit from them
2) combine the error handling for flushing packets into one efficient loop
3) have the user-provided callbacks interoperate with streams without issue

I'll untangle this, starting probably with point 2